sympy/solvers/solveset.py

```diff
@@ -1121,3 +1125,823 @@ def linsolve(system, *symbols):
+
+    >>> substitution([x + y], [x], [{y: 1}], [y], set([x + 1]), [y, x])
+    EmptySet()
+    >>> substitution([x + y], [x], [{y: 1}], [y], set([x - 1]), [y, x])
+    {(1, -1)}
+    >>> substitution([x + y - 1, y - x**2 + 5], [x, y])
+    {(-3, 4), (2, -1)}
+
+    * Returns both real and complex solution
+
+    >>> x, y, z = symbols('x, y, z')
+    >>> from sympy import exp, sin
+    >>> substitution([exp(x) - sin(y), y**2 - 4], [x, y])
+    {(log(sin(2)), 2), (ImageSet(Lambda(_n, I*(2*_n*pi + pi) +
+        log(sin(2))), Integers()), -2), (ImageSet(Lambda(_n, 2*_n*I*pi +
+        Mod(log(sin(2)), 2*I*pi)), Integers()), 2)}
+
+    >>> eqs = [z**2 + exp(2*x) - sin(y), -3 + exp(-y)]
+    >>> substitution(eqs, [y, z])
+    {(-log(3), -sqrt(-exp(2*x) - sin(log(3)))),
+    (-log(3), sqrt(-exp(2*x) - sin(log(3)))),
+    (ImageSet(Lambda(_n, 2*_n*I*pi + Mod(-log(3), 2*I*pi)), Integers()),
+    ImageSet(Lambda(_n, -sqrt(-exp(2*x) + sin(2*_n*I*pi +
+    Mod(-log(3), 2*I*pi)))), Integers())),
+    (ImageSet(Lambda(_n, 2*_n*I*pi + Mod(-log(3), 2*I*pi)), Integers()),
+    ImageSet(Lambda(_n, sqrt(-exp(2*x) + sin(2*_n*I*pi +
+        Mod(-log(3), 2*I*pi)))), Integers()))}
+
+    """
+
+    from sympy import Complement
+    from sympy.core.compatibility import is_sequence
+
+    if not system:
+        return S.EmptySet
+
+    if not symbols:
+        raise ValueError('Symbols must be given, for which solution of the '
+                         'system is to be found.')
+
+    try:
+        sym = symbols[0].is_Symbol
+    except AttributeError:
+        sym = False
+
+    if not sym:
+        raise ValueError('Symbols or iterable of symbols must be given as '
+                         'second argument, not type \
+            %s: %s' % (type(symbols[0]), symbols[0]))
+
+    if not is_sequence(symbols):
+                raise TypeError(
+                    'syms should be given as a sequence, e.g. a list')
+
+    # By default `all_symbols` will be same as `symbols`
+    if all_symbols is None:
+        all_symbols = symbols
+
+    old_result = result
+    # storing complements and intersection for particular symbol
+    complements = {}
+    intersections = {}
+
+    # when total_solveset_call is equals to total_conditionset
+    # means solvest failed to solve all the eq.
+    total_conditionset = -1
+    total_solveset_call = -1
+
+    def _unsolved_syms(eq, sort=False):
+        """Returns the unsolved symbol present
+        in the equation `eq`.
+        """
+        free = eq.free_symbols
+        unsolved = (free - set(known_symbols)) & set(all_symbols)
+        if sort:
+            unsolved = list(unsolved)
+            unsolved.sort(key=default_sort_key)
+        return unsolved
+
+    # sort such that equation with the fewest potential symbols is first.
+    # means eq with less variable first
+    eqs_in_better_order = list(
+        ordered(system, lambda _: len(_unsolved_syms(_))))
+
+    def _return_conditionset():
```

Comment:
There is no point in defining a function which uses the local variable of another function as it would not be useful outside the scope.